superqt/utils/_ensure_thread.py

```diff
@@ -102,33 +114,22 @@ def _func(self, *args, **kwargs):
 
 
 def _run_in_thread(
-    func: Callable, thread: QThread, await_return: bool, timeout: int, *args, **kwargs
-):
+    func: Callable[..., T],
+    thread: QThread,
+    await_return: bool,
+    timeout: int,
+    *args,
+    **kwargs,
+) -> Union[Future[T], T]:
+    future: Future[T] = Future()
     if thread is QThread.currentThread():
+        result = func(*args, **kwargs)
         if not await_return:
-            func(*args, **kwargs)
-            return
-        return func(*args, **kwargs)
-    f = CallCallable(func, *args, **kwargs)
+            future.set_result(result)
+            return future
+        return result
+    f = _CallCallable(func, *args, **kwargs)
     f.moveToThread(thread)
-    if not await_return:
-        QMetaObject.invokeMethod(f, "call", Qt.ConnectionType.QueuedConnection)
-        return
-
-    res = []
-
-    def set_res(data):
-        res.append(data)
-
-    f.finished.connect(set_res)
-    timer = QTimer()
-    timer.setSingleShot(True)
-    loop = QEventLoop()
-    f.finished.connect(loop.quit)
-    timer.timeout.connect(loop.quit)
-    timer.start(timeout)
-    QMetaObject.invokeMethod(f, "call", Qt.ConnectionType.QueuedConnection)
-    loop.exec_()
-    if len(res) == 0:
-        raise TimeoutError("Not recived value")
-    return res[0]
+    f.finished.connect(future.set_result, Qt.ConnectionType.DirectConnection)
```

Comment:
please pay close attention here...   it doesn't work unless you use `DirectConnection`.  I think it's safe here, but would like for you to confirm

Reply:
it looks safe